src/Rector/AbstractRector.php

```diff
@@ -182,6 +194,7 @@ public function beforeTraverse(array $nodes): ?array
     }
 
     /**
+     * @phpstan-param TNodeType $node
```

Comment:
This one is wrong. `enterNode` is called with any node (as it has not applied the node type matching yet, as that is done inside the method)

Reply:
You're right, `$node` is a `TNodeType` after calling `isMatchingNodeType()`, but...

```
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Rector/AbstractRector.php                                                                                                                                                    
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  214    Parameter #1 $node of method Rector\Core\Contract\Rector\PhpRectorInterface<TNodeType of PhpParser\Node>::refactor() expects TNodeType of PhpParser\Node, PhpParser\Node given.  
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
```

That's because PHPStan does not know that after calling `isMatchingNodeType()` variable `$node` is a `TNodeType`, not `Node` 🤔